sklearn/metric_learning/nca.py

```diff
@@ -0,0 +1,277 @@
+from ..base import BaseEstimator, TransformerMixin
+from ..utils.validation import check_is_fitted
+
+import numpy as np
+import scipy as sp
+import scipy.misc
+import scipy.optimize
+import time
+
+import sys
+
+
+def nca_vectorized_oracle(L, X, y, n_components, loss, outer, threshold=0.0):
+    n_samples, n_features = X.shape
+
+    L = L.reshape((n_components, n_features))
+
+    Lx = np.dot(X, L.T)  # n_samples x n_components
+    assert Lx.shape == (n_samples, n_components)
+
+    A = Lx[np.newaxis, :, :] - Lx[:, np.newaxis, :]  # n_samples x n_samples x n_components
+    assert A.shape == (n_samples, n_samples, n_components)
+
+    logp = -np.einsum("ijk,ijk->ij", A, A)  # n_samples x n_samples
+    np.fill_diagonal(logp, -np.inf)
+    logp -= sp.misc.logsumexp(logp, axis=1)[:, np.newaxis]
+    assert logp.shape == (n_samples, n_samples)
+
+    p = np.exp(logp)  # n_samples x n_samples
+    assert p.shape == (n_samples, n_samples)
+
+    class_neighbours = y[:, np.newaxis] == y[np.newaxis, :]
+
+    p_i = np.einsum("ij,ij->i", p, class_neighbours)
+
+    proba = np.copy(p)
+    if loss == 'l1':
+        proba *= p_i[:, np.newaxis]
+    mask = proba > threshold
+    proba *= mask
+
+    grad = np.einsum("ij,ijkl", proba, outer)
+    assert grad.shape == (n_features, n_features)
+
+    neighbours_proba = class_neighbours * p
+    if loss == 'kl':
+        nonzero = p_i > 1e-10
+        neighbours_proba[nonzero, :] /= p_i[nonzero, np.newaxis]
+
+    neighbours_mask = neighbours_proba > threshold
+    neighbours_proba *= neighbours_mask
+
+    neighbours_term = np.einsum("ij,ijkl", neighbours_proba, outer)
+
+    assert neighbours_term.shape == (n_features, n_features)
+    grad -= neighbours_term
+
+    function_value = 0
+    if loss == 'l1':
+        function_value = p_i.sum()
+    elif loss == 'kl':
+        function_value = np.log(p_i).sum()
+
+    grad = 2 * np.dot(L, grad)
+
+    return [-function_value, -grad.flatten()]
+
+
+def nca_semivectorized_oracle(L, X, y, n_components, loss, threshold=0.0):
+    n_samples, n_features = X.shape
+    L = L.reshape((n_components, n_features))
+
+    Lx = np.dot(X, L.T)  # n_samples x n_components
+    assert Lx.shape == (n_samples, n_components)
+
+    grad = np.zeros((n_features, n_features))
+    function_value = 0
+
+    distances = np.zeros((n_samples, n_samples))
+    for i in range(n_samples):
```

Comment:
Could you check this full for-loop against `sklearn.metrics.euclidean_distances(..., square=True)` (https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/metrics/pairwise.py#L136) or make use of the tricks implemented there, which include https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/utils/extmath.py#L57 ? Although it is not informed of the symmetry of the result, it maximally exploits the reusability of certain calculations, and may still be competitive.